File: src/tpuff/commands/env.py
"""Environment management commands for tpuff."""

import click
from rich.console import Console
from rich.table import Table

from tpuff.config import add_env, get_active_env, list_envs, remove_env, set_active
from tpuff.utils.output import is_plain, print_table_plain
from tpuff.utils.regions import DEFAULT_REGION, TURBOPUFFER_REGIONS

console = Console()


def mask_key(api_key: str) -> str:
    """Mask an API key, showing only the first 8 characters."""
    if len(api_key) <= 8:
        return api_key
    return api_key[:8] + "..."


@click.group("env", context_settings={"help_option_names": ["-h", "--help"]})
def env():
    """Manage tpuff environments."""
    pass


@env.command("add")
@click.argument("name")
@click.pass_context
def env_add(ctx: click.Context, name: str) -> None:
    """Add a new environment (interactive setup)."""
    api_key = click.prompt("API key", hide_input=True)
    region = click.prompt(
        "Region",
        type=click.Choice(TURBOPUFFER_REGIONS, case_sensitive=False),
        default=DEFAULT_REGION,
    )
    base_url = click.prompt("Base URL (optional, press Enter to skip)", default="")

    add_env(name, api_key, region, base_url)
    click.echo(f"Environment '{name}' added.")

    # Show if it was set as active
    active = get_active_env()
    if active and active[0] == name:
        click.echo("Set as active environment.")


@env.command("use")
@click.argument("name")
def env_use(name: str) -> None:
    """Switch active environment."""
    set_active(name)
    click.echo(f"Switched to environment '{name}'.")


@env.command("list")
@click.pass_context
def env_list(ctx: click.Context) -> None:
    """List all environments."""
    envs = list_envs()
    if not envs:
        click.echo("No environments configured. Run 'tpuff env add <name>' to add one.")
        return

    plain = is_plain(ctx)
    headers = ["", "Name", "Region", "API Key"]
    rows = []
    for name, env_dict, is_active in envs:
        marker = "*" if is_active else ""
        rows.append([
            marker,
            name,
            env_dict.get("region", DEFAULT_REGION),
            mask_key(env_dict.get("api_key", "")),
        ])

    if plain:
        print_table_plain(headers, rows)
    else:
        table = Table(show_header=True, header_style="cyan")
        for h in headers:
            table.add_column(h)
        for r in rows:
            # Highlight active env
            if r[0] == "*":
                table.add_row(
                    "[green]*[/green]",
                    f"[bold]{r[1]}[/bold]",
                    r[2],
                    f"[dim]{r[3]}[/dim]",
                )
            else:
                table.add_row("", r[1], r[2], f"[dim]{r[3]}[/dim]")
        console.print(table)


# Alias
env.add_command(env_list, name="ls")


@env.command("rm")
@click.argument("name")
@click.confirmation_option(prompt="Are you sure you want to remove this environment?")
def env_rm(name: str) -> None:
    """Remove an environment."""
    remove_env(name)
    click.echo(f"Environment '{name}' removed.")


@env.command("show")
def env_show() -> None:
    """Show the current active environment."""
    active = get_active_env()
    if not active:
        click.echo("No active environment. Run 'tpuff env add <name>' to add one.")
        return

    name, env_dict = active
    click.echo(f"Active environment: {name}")
    click.echo(f"  Region:   {env_dict.get('region', DEFAULT_REGION)}")
    click.echo(f"  API Key:  {mask_key(env_dict.get('api_key', ''))}")
    base_url = env_dict.get("base_url", "")
    if base_url:
        click.echo(f"  Base URL: {base_url}")

File: cmd/env.go
package cmd

import (
	"bufio"
	"fmt"
	"os"
	"strings"

	"github.com/hev/tpuff/internal/config"
	"github.com/hev/tpuff/internal/output"
	"github.com/hev/tpuff/internal/regions"
	"github.com/spf13/cobra"
)

var envCmd = &cobra.Command{
	Use:   "env",
	Short: "Manage tpuff environments",
}

var envAddCmd = &cobra.Command{
	Use:   "add NAME",
	Short: "Add a new environment (interactive setup)",
	Args:  cobra.ExactArgs(1),
	RunE:  runEnvAdd,
}

var envUseCmd = &cobra.Command{
	Use:   "use NAME",
	Short: "Switch active environment",
	Args:  cobra.ExactArgs(1),
	RunE:  runEnvUse,
}

var envListCmd = &cobra.Command{
	Use:     "list",
	Aliases: []string{"ls"},
	Short:   "List all environments",
	RunE:    runEnvList,
}

var envRmCmd = &cobra.Command{
	Use:   "rm NAME",
	Short: "Remove an environment",
	Args:  cobra.ExactArgs(1),
	RunE:  runEnvRm,
}

var envShowCmd = &cobra.Command{
	Use:   "show",
	Short: "Show the current active environment",
	RunE:  runEnvShow,
}

var envSetContentCmd = &cobra.Command{
	Use:   "set-content FIELD",
	Short: "Set the content field shown in the browser preview",
	Long: `Set which document field is displayed as the preview in the interactive browser.

Without --namespace, sets the default for the active environment.
With --namespace, sets an override for that specific namespace.

Use an empty string "" to clear the setting.`,
	Args: cobra.ExactArgs(1),
	RunE: runEnvSetContent,
}

func init() {
	envSetContentCmd.Flags().StringP("namespace", "n", "", "Set content field for a specific namespace")

	envCmd.AddCommand(envAddCmd)
	envCmd.AddCommand(envUseCmd)
	envCmd.AddCommand(envListCmd)
	envCmd.AddCommand(envRmCmd)
	envCmd.AddCommand(envShowCmd)
	envCmd.AddCommand(envSetContentCmd)
	rootCmd.AddCommand(envCmd)
}

func runEnvAdd(cmd *cobra.Command, args []string) error {
	name := args[0]
	reader := bufio.NewReader(os.Stdin)

	fmt.Print("API key: ")
	apiKey, _ := reader.ReadString('\n')
	apiKey = strings.TrimSpace(apiKey)

	fmt.Printf("Region [%s]: ", regions.DefaultRegion)
	region, _ := reader.ReadString('\n')
	region = strings.TrimSpace(region)
	if region == "" {
		region = regions.DefaultRegion
	}

	if !regions.IsValid(region) {
		fmt.Fprintf(os.Stderr, "Error: Unknown region '%s'. Valid regions:\n", region)
		for _, r := range regions.TurbopufferRegions {
			fmt.Fprintf(os.Stderr, "  %s\n", r)
		}
		os.Exit(1)
	}

	fmt.Print("Base URL (optional, press Enter to skip): ")
	baseURL, _ := reader.ReadString('\n')
	baseURL = strings.TrimSpace(baseURL)

	if err := config.AddEnv(name, apiKey, region, baseURL); err != nil {
		fmt.Fprintf(os.Stderr, "Error: %s\n", err)
		os.Exit(1)
	}

	fmt.Printf("Environment '%s' added.\n", name)

	_, _, ok := config.GetActiveEnv()
	if ok {
		activeName, _, _ := config.GetActiveEnv()
		if activeName == name {
			fmt.Println("Set as active environment.")
		}
	}
	return nil
}

func runEnvUse(cmd *cobra.Command, args []string) error {
	name := args[0]
	if err := config.SetActive(name); err != nil {
		fmt.Fprintf(os.Stderr, "Error: %s\n", err)
		os.Exit(1)
	}
	fmt.Printf("Switched to environment '%s'.\n", name)
	return nil
}

func runEnvList(cmd *cobra.Command, args []string) error {
	envs := config.ListEnvs()
	if len(envs) == 0 {
		fmt.Println("No environments configured. Run 'tpuff env add <name>' to add one.")
		return nil
	}

	headers := []string{"", "Name", "Region", "API Key"}
	var rows [][]string
	for _, e := range envs {
		marker := ""
		if e.IsActive {
			marker = "*"
		}
		rows = append(rows, []string{
			marker,
			e.Name,
			e.Config.Region,
			config.MaskKey(e.Config.APIKey),
		})
	}

	output.PrintTable(headers, rows)
	return nil
}

func runEnvRm(cmd *cobra.Command, args []string) error {
	name := args[0]
	if !confirmPrompt(fmt.Sprintf("Are you sure you want to remove environment '%s'?", name)) {
		fmt.Println("Aborted")
		return nil
	}
	if err := config.RemoveEnv(name); err != nil {
		fmt.Fprintf(os.Stderr, "Error: %s\n", err)
		os.Exit(1)
	}
	fmt.Printf("Environment '%s' removed.\n", name)
	return nil
}

func runEnvShow(cmd *cobra.Command, args []string) error {
	name, env, ok := config.GetActiveEnv()
	if !ok {
		fmt.Println("No active environment. Run 'tpuff env add <name>' to add one.")
		return nil
	}

	fmt.Printf("Active environment: %s\n", name)
	fmt.Printf("  Region:   %s\n", env.Region)
	fmt.Printf("  API Key:  %s\n", config.MaskKey(env.APIKey))
	if env.BaseURL != "" {
		fmt.Printf("  Base URL: %s\n", env.BaseURL)
	}
	if env.ContentField != "" {
		fmt.Printf("  Content:  %s\n", env.ContentField)
	}
	if len(env.ContentFields) > 0 {
		fmt.Println("  Content overrides:")
		for ns, field := range env.ContentFields {
			fmt.Printf("    %s → %s\n", ns, field)
		}
	}
	return nil
}

func runEnvSetContent(cmd *cobra.Command, args []string) error {
	field := args[0]
	namespace, _ := cmd.Flags().GetString("namespace")

	if err := config.SetContentField(field, namespace); err != nil {
		fmt.Fprintf(os.Stderr, "Error: %s\n", err)
		os.Exit(1)
	}

	if namespace != "" {
		if field == "" {
			fmt.Printf("Cleared content field for namespace '%s'.\n", namespace)
		} else {
			fmt.Printf("Content field for namespace '%s' set to '%s'.\n", namespace, field)
		}
	} else {
		if field == "" {
			fmt.Println("Cleared default content field.")
		} else {
			fmt.Printf("Default content field set to '%s'.\n", field)
		}
	}
	return nil
}
